### DISCUSSION TITLE
Enhancements to SetTitle

### DISCUSSION THREAD

**Discussion by markcellis** [elgatosf/elgato#17] [Ideas]
It would be really great if we could have a bit more control over the text being displayed on a key using the SetTitle API message. At the moment the font, size, and color are all controlled on a context using the Stream Deck software. To the best of my knowledge, there is no way to change it via an SDK message. All you can change is the text string with SetTitle.

It would be great if we had more control over the text being displayed, size, color, and alignment on a line-by-line base. Font face would also be great but I realize that might be challenging given the multiple operating systems. Maybe we need something like the Layouts we have on the Stream Deck Plus encoder panels,

**Comment by ChekTek** [comment]
Hi!

This is great feedback, I will create a feature request internally. Although the fine grain control you are looking for using the `setTitle` API is not available at the moment, you can set some of the default styling for each action's state in the manifest. 

https://developer.elgato.com/documentation/stream-deck/sdk/manifest/#states

**Reply by markcellis** [reply]
Yes, I'm aware that I can do it in the manifest but I need to do it from the plugin based on dynamic data.

**Comment by UweSchwo** [comment]
Great idea!
Hope we can finally get some "format at once" plugins this way. This means: Mark e.g. 8 buttons and change font size or font color with only one command.  For me this is a very common use case.
Moreover it would be nice to have an option to change the background picture for multiple buttons at once. But I know, it's a different problem.😏

**Comment by smitmartijn** [comment]
I'd like to +1 this one. Reason: I'm building a plugin that shows statistics (numbers) on a key and I've got an optional header text. With the header text, I'm align the text to the top and using newlines to put the header on the top of the key, and the statistic in the middle. Without the header text, it's not possible to align the statistic in the center with line breaks alone, so it would be awesome to be able to tell it to align in the center for that case.

![2023-09-08_11-19-39](https://github.com/elgatosf/elgato/assets/6500836/187a3188-b775-4ed3-a174-ee181bebe68c)

**Comment by CrazyCoder** [comment]
Another vote for this, with a concrete case.

I maintain a plugin that shows numeric counters on keys. While rebuilding a setup I found keys configured years ago whose profile carries `FontSize: 24`. The runtime still renders them fine, but the current UI slider stops at 18, so those keys cannot be recreated through the UI anymore, and a plugin has no API to do it either. The app clearly supports larger titles; there is just no way left to ask for them.

The workaround everyone lands on is drawing the text into the key image, and that path is bumpier than it looks. SVG `<text>` elements don't rasterize, so text has to be converted to outline paths first (I bundle a font and use opentype.js). Raw SVG strings were rejected in my testing on Stream Deck 7.5 for Windows despite being listed as supported in the `setImage` docs, so the image has to go in as a base64 data URI. And once a user assigns a custom icon to a key, `setImage` is ignored entirely, so the plugin also has to grow its own icon setting and composite the icon into the image itself. To keep the result visually consistent with native titles I read `titleParameters` and resolve the selected family against the OS font table at runtime. That is a lot of machinery to reimplement something the app already does well.

Runtime control of title parameters, size at minimum, would remove the need for all of it. Failing that, raising the slider cap would already help, since the renderer handles larger values today. Happy to share repro details for the `setImage` quirks if useful.

**Comment by UweSchwo** [comment]
I don't understand the 18pt limit too. But I worked around by directly editing the json file. Meanwhile I've buttons for fast finding the right json file and the right position within for manipulating things like font size.
